providers/os/resources/cloud/instance_metadata.go

```diff
@@ -52,3 +57,82 @@ func (m InstanceMetadata) PrivateIP() string {
 	}
 	return ""
 }
+
+// AddOrUpdatePublicIP adds or updates one or many Ipv4Addresses
+func (m *InstanceMetadata) AddOrUpdatePublicIP(ips ...Ipv4Address) {
+	if m.PublicIpv4 == nil {
+		m.PublicIpv4 = make([]Ipv4Address, 0)
+	}
+
+	for _, ip := range ips {
+		index := m.findPublicIP(ip.IP)
+		if index < 0 {
+			// not found, add
+			log.Trace().Str("ip", ip.IP).Msg("os.cloud.metadata> add public ip")
+			m.PublicIpv4 = append(m.PublicIpv4, ip)
+			continue
+		}
+
+		// found, update
+		log.Trace().Str("ip", ip.IP).Msg("os.cloud.metadata> update public ip")
+		m.mergePublicIP(index, ip)
+	}
+}
+func (m *InstanceMetadata) mergePublicIP(index int, ip Ipv4Address) {
+	merged := mergeIPs(m.PublicIpv4[index], ip)
+	m.PublicIpv4[index] = merged
+}
+func (m *InstanceMetadata) findPublicIP(ip string) int {
+	return slices.IndexFunc(m.PublicIpv4, func(address Ipv4Address) bool {
+		return address.IP == ip
+	})
+}
+
+// AddOrUpdatePrivateIP adds or updates one or many Ipv4Addresses
+func (m *InstanceMetadata) AddOrUpdatePrivateIP(ips ...Ipv4Address) {
```

Comment:
whats the use case in which we get the same ip twice so it needs to be updated? i'd expect the metadata to give us the IP addresses just once

Reply:
After doing extensive research, I discovered that the public IP address may come from two different place within the IMDS, either the `instance/` metadata or the `loadbalander/` one.

Docs:
https://learn.microsoft.com/en-us/azure/virtual-machines/instance-metadata-service?tabs=windows#endpoint-categories

So we don't know where will it come from, but we check both places looking. The question is, what do we do if it is present in both places? I say we merge the information in the best possible way, which is this implementation, but I am open to suggestions.

Reply:
if the instance is behind a LB, will the instance endpoint still give us back the IP info?
Maybe it's best to just define a precedence order here and use whichever one has higher prio?

This code is also fine, I just want to make sure there isnt an easier way to handle this case

Reply:
The precedence comes from who calls first this function, we prioritize the information from the load balancer, if it exits. Then look at the instance metadata:

https://github.com/mondoohq/cnquery/blob/8b43a91b47fdcff961398a27d76deedb6524563b/providers/os/resources/cloud/azure.go#L53-L115

I can move the logic from here to there 👆🏽  if that makes things easier.